File: Algo/Leetcode/039CombinationSum.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> r;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        int n = candidates.size();
        if(n==0) return res;
        dfs(candidates, target, 0, 0);
        return res;
    }

    void dfs(vector<int>& candidates, int target, int sum, int start){
        if(sum == target){
            res.push_back(r);
            return;
        }

        if(sum > target){  // need this to stop the search. Since all candidates are positive.
            return;
        }

        for(int i=start; i<candidates.size(); i++){
            r.push_back(candidates[i]);
            sum += candidates[i];
            dfs(candidates, target, sum, i); //repeatable
            r.pop_back();
            sum -= candidates[i];
        }
    }
};

File: Algo/Leetcode/039CombinationSum.py
class Solution(object):

    def __init__(self):
        self.res = []
        self.path = []
        self.sum = 0

    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        n = len(candidates)
        self.dfs(candidates, 0, n, target)

        return self.res

    def dfs(self, candidates, i, n, target):
        if self.sum == target:
            self.res.append(copy.deepcopy(self.path))
            return

        if i >= n or self.sum > target:
            return

        for j in range(i, n):
            self.sum += candidates[j]
            self.path.append(candidates[j])
            self.dfs(candidates, j, n, target)
            self.path.pop()
            self.sum -= candidates[j]